File: backend/src/main/java/com/internship/tool/service/ToolService.java
package com.internship.tool.service;

import com.internship.tool.dto.ToolRequest;
import com.internship.tool.dto.ToolResponse;
import org.springframework.data.domain.Page;

import java.util.List;

public interface ToolService {

    ToolResponse createTool(ToolRequest request);

    List<ToolResponse> getAllTools();

    ToolResponse getToolById(Long id);

    ToolResponse updateTool(Long id, ToolRequest request);

    void deleteTool(Long id);

    Page<ToolResponse> searchTools(
            String name,
            String category,
            int page,
            int size,
            String sortBy,
            String direction
    );
}

File: frontend/src/services/toolService.js
const API_BASE_URL = "http://localhost:8081/api/tools";

export const getAllTools = async () => {
  const response = await fetch(API_BASE_URL);
  return response.json();
};

export const searchTools = async ({
  name = "",
  category = "",
  page = 0,
  size = 5,
  sortBy = "id",
  direction = "asc",
}) => {
  const params = new URLSearchParams();

  if (name) params.append("name", name);
  if (category) params.append("category", category);

  params.append("page", page);
  params.append("size", size);
  params.append("sortBy", sortBy);
  params.append("direction", direction);

  const response = await fetch(`${API_BASE_URL}/search?${params.toString()}`);
  return response.json();
};

export const createTool = async (toolData) => {
  const response = await fetch(API_BASE_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(toolData),
  });

  return response.json();
};

export const updateTool = async (id, toolData) => {
  const response = await fetch(`${API_BASE_URL}/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(toolData),
  });

  return response.json();
};

export const deleteTool = async (id) => {
  const response = await fetch(`${API_BASE_URL}/${id}`, {
    method: "DELETE",
  });

  return response.text();
};